src/Streamly/Internal/Unicode/Normalize.hs

```diff
@@ -0,0 +1,273 @@
+        return $ Yield ch (YieldCharList chs ns)
+    sstep gst (ReadInputChar rebuf st) = do
+        res <- step gst st
+        return
+          $ Skip
+          $ case res of
+                Yield ch st1
+                    | UC.isHangul ch ->
+                        if decomposeHangul
+                        then YieldCharList rebuf (IsHangul ch st1)
+                        else YieldCharList
+                                 (rebuf ++ [ch])
+                                 (ReadInputChar [] st1)
+                    | UC.isDecomposable mode ch ->
+                        IsDecomposable (UC.decompose mode ch) rebuf st1
+                    | otherwise ->
+                        if UC.isCombining ch
+                        then ReadInputChar (insertIntoReBuf ch rebuf) st1
+                        else YieldCharList
+                                 (rebuf ++ [ch])
+                                 (ReadInputChar [] st1)
+                Skip st1 -> ReadInputChar rebuf st1
+                Stop -> YieldCharList rebuf DecomposeStop
+    sstep _ (IsHangul ch st) =
+        return
+          $ Skip
+          $ let (l, v, t) = UC.decomposeHangul ch
+             in if t == chr UC.jamoTFirst
+                then YieldCharList [l, v] (ReadInputChar [] st)
+                else YieldCharList [l, v, t] (ReadInputChar [] st)
+    sstep _ (IsDecomposable [] rebuf st) =
+        return $ Skip $ ReadInputChar rebuf st
+    sstep _ (IsDecomposable (ch:chs) rebuf st)
+        | UC.isDecomposable mode ch =
+            return
+              $ Skip $ IsDecomposable (UC.decompose mode ch ++ chs) rebuf st
+        | otherwise =
+            return
+              $ Skip
+              $ if UC.isCombining ch
+                then IsDecomposable chs (insertIntoReBuf ch rebuf) st
+                else YieldCharList (rebuf ++ [ch]) (IsDecomposable chs [] st)
+    sstep _ DecomposeStop = return Stop
+
+{-# INLINE decomposeD #-}
+decomposeD :: Monad m => DecomposeMode -> Stream m Char -> Stream m Char
+decomposeD = partialDecomposeD True
+
+-- Hold an L to wait for V, hold an LV to wait for T.
+data JamoBuf
+    = Jamo !Char -- Jamo L, V or T
+    | Hangul !Char -- Hangul Syllable LV or LVT
+    | HangulLV !Char
+
+{-# INLINE fromJamoBuf #-}
+fromJamoBuf :: JamoBuf -> Char
+fromJamoBuf (Jamo ch) = ch
+fromJamoBuf (Hangul ch) = ch
+fromJamoBuf (HangulLV ch) = ch
+
+
+-- {-# ANN type ComposeState Fuse #-}
+data ComposeState st
+    = YieldChar Char (ComposeState st)
+    | YieldList [Char] (ComposeState st)
+    | ComposeNone st
+    | ComposeReg Int [Char] st
+    | ComposeJamo JamoBuf st
+    | ComposeStop
+
+-- Assumes every character except hangul characters are fully decomposed and the
+-- combining characters are reordered. Hangul characters may or may not be
+-- decomposed.
+{-# INLINE_EARLY partialComposeD #-}
+partialComposeD :: Monad m => Stream m Char -> Stream m Char
+partialComposeD (Stream step state) = Stream step' (ComposeNone state)
```

Comment:
call this `composeD`.

Reply:
This is kind of partially composing. This does not compose a semi-composed stream. It shouldn't be used directly.